### ISSUE TITLE
Allow characters/custom functions to set the values

### ISSUE THREAD

**Issue by Princezhm** [s-yadav/react-number-format#196]
I was wondering if there is a way to set a pre evaluation function to evaluate some inputs, for example I was thinking on an user which enters "1K" and if it is a valid entry a function will be able to return the actual value (1000).

if null is returned, it is not valid or it doesnt have a valid mapping.

```javascript
function preEvaluator(value){
    if(value === "1k") return 1000;

    return null
}
......
......

<NumberFormat preEvaluate={preEvaluator} ..... />
```

is there a way to accomplish this now? because I don't see a way to do it.

**Comment by awesor** [comment]
I was looking for the exact same thing.  Unfortunately, I don't think that any alpha characters are allowed, unless they're passed via the suffix.  Even then, the user cannot choose their own suffix ("K", "M" - millions, "B"- billions).

Definitely would love to see alpha-chars supported!

**Comment by s-yadav** [comment]
@Princezhm @awesor preEvaluate can be a option, where we can allow preEvaluate to transform number before NumberFormat do any transformation on it. The only problem with preEvaluate is how we will calculate correct caret position after it.

I can think of this as an experimental props with limited use case. 

Well, I am open for discussion on this feature on what all cases can come with this prop and what we can support.

**Comment by Princezhm** [comment]
hi @s-yadav I want to contribute to your repo, so I started to see what we can do, for example we got this options, maybe @awesor can add some cases:

- have a super smart input to modify the different inputs, for example typing `"x"` on top of `1.233|452`, which will result on `1.233x|452`, but what can we do with this? I don't get a case for this.
- just have let the user to validate the key that was pressed and return null if nothing needs to be done, in any other case return a number/float to multiply the current value to the new one. for example if the user has `1.22` and in our `customFunction` we receive `k` we can return `1024` (obviously this is up to the user) and therefore we will multiply `1.22 * 1024 =1249.28 ` and we apply the correct formatting to the `1249.28` number, which is the final result.  one thing to notice with this is that if we do it this way, only allowing the final key to be the one to use as a precalculator, we don't need to worry about the caret position (correct me if I'm wrong), because it will always end at the end of the value, in the example it goes from `1.22|` to `1249.28|`

so, the custom transformation function could be something like:

```javascript
/**
   *
   * @param { string } key: key that was pressed by the user, this function will be called only on special keys.
   * @returns { Number / Null } Number to multiply the value or null if the user doesnt allow the character
   */
  customTransformation(key) {
    switch (key) {
      case "k":
      case "K":
        return 1000;
      case "m":
      case "M":
        return 1000000;
      case "n":
      case "N":
        return 0.01; //instead of creating a division, we say that we need floating points to multiply ( value / 100 === value * 0.01)
      default:
        return 1; // just if the user wants to handle what happens if nothing passed
    }
    return null; // nothing to do, then we stop handling the event and we remove the character.
  }
.......
render() {
    return (
      <NumberFormat
           customCalculator={this.customTransformation}
          />
    );
  }
```

now there are special cases, like the user pressing the backspace or special keys, we should have a way to avoid to send this keys to the user, would be good to ask the user to send a map of keys? with that we can avoid to use the custom function. something like this:

```javascript
render() {
    const keysMap = {
      k: 1000,
      m: 1000000,
      n: 0.01
    }
    return (
      <NumberFormat
           customKeys={keysMap}
          />
    );
  }
```

what do you think? I was thinking while I was writing, so excuse me if I forgot something.

**Comment by Princezhm** [comment]
but that comparison will be done on the  `OnKeyDown` method or in the `OnChange` method?  where do you suggest to have a functionality like that?

**Comment by s-yadav** [comment]
OnKeyDown is not very reliable across devices, for android chrome key codes are incorrect. Not sure if they have fixed it. I had to do lot of patch because of it.

**Comment by Princezhm** [comment]
so the best solution would be to change the values on the `OnChange` method before setting the state and returning the final value to the user, correct? on these days I'll start looking at it and see if we can get something usefull

**Comment by s-yadav** [comment]
Yes onChange is the safest place to do this. Possibly after this https://github.com/s-yadav/react-number-format/blob/master/src/number_format.js#L637. After correcting the value.
